java-checks/src/main/java/org/sonar/java/checks/ArrayCopyLoopCheck.java

```diff
@@ -0,0 +1,309 @@
+        ArrayAccessExpressionTree dst = (ArrayAccessExpressionTree) lhs;
+        return isIdentifier(src.expression()) && useCounter(src, ctx) && isIdentifier(dst.expression()) && useCounter(dst, ctx);
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArrayToListCopy(ExpressionTree expression, Context ctx) {
+    if (expression.is(Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree invocation = (MethodInvocationTree) expression;
+      if (invocation.methodSelect().is(Kind.MEMBER_SELECT) && invocation.arguments().size() == 1) {
+        MemberSelectExpressionTree select = (MemberSelectExpressionTree) invocation.methodSelect();
+        ExpressionTree object = select.expression();
+        IdentifierTree method = select.identifier();
+        if (isIdentifier(object) && isCollection(object) && "add".equals(method.symbol().name())) {
+          ExpressionTree argument = invocation.arguments().get(0);
+          if (argument.is(Kind.ARRAY_ACCESS_EXPRESSION)) {
+            ArrayAccessExpressionTree access = (ArrayAccessExpressionTree) argument;
+            return isIdentifier(access.expression()) && useCounter(access, ctx);
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArrayToListCopy(ExpressionTree expression, VariableTree iterated) {
+    if (expression.is(Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree invocation = (MethodInvocationTree) expression;
+      if ("add".equals(invocation.symbol().name()) && invocation.arguments().size() == 1) {
+        expression = invocation.methodSelect();
+        if (expression.is(Kind.MEMBER_SELECT)) {
+          MemberSelectExpressionTree select = (MemberSelectExpressionTree) expression;
+          ExpressionTree object = select.expression();
+          if (isIdentifier(object) && isCollection(object)) {
+            ExpressionTree argument = invocation.arguments().get(0);
+            IdentifierTree identifier = getIdentifier(argument);
+            return identifier != null && identifier.symbol().equals(iterated.symbol());
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArray(ExpressionTree tree) {
+    return tree.symbolType().isArray();
+  }
+
+  private static boolean isCollection(ExpressionTree tree) {
+    return tree.symbolType().isSubtypeOf("java.util.Collection");
+  }
+
+  private static boolean isIdentifier(ExpressionTree tree) {
+    return tree.is(Kind.IDENTIFIER);
+  }
+
+  private static IdentifierTree getIdentifier(ExpressionTree tree) {
+    if (isIdentifier(tree)) {
+      return (IdentifierTree) tree;
+    }
+    return null;
+  }
+
+  private static boolean isCounter(ExpressionTree tree, Context ctx) {
+    IdentifierTree identifier = getIdentifier(tree);
+    return identifier != null && identifier.symbol().equals(ctx.getCounter());
+  }
+
+  private static boolean isInteger(ExpressionTree tree) {
```

Comment:
I would suggest renaming this to `isIntegerType`, to make it less ambiguous (from the name I was thinking that you are testing the expression to be integer literal)